xc/domain.ml

```diff
@@ -287,49 +287,83 @@ let wait_xen_free_mem ~xc ?(maximum_wait_time_seconds = 64) required_memory_kib
+  let open Xenctrl in
   let host_info = Xenctrl.physinfo xc in
-  let host_has_hap = Xenctrl.(List.mem CAP_HAP host_info.capabilities) in
-  debug "Host Hardware Assisted Paging is %s"
-    (if host_has_hap then "available" else "N/A") ;
-  let flags =
-    match List.assoc_opt "hap" vm_info.platformdata with
-    | Some "true" when vm_info.hvm ->
-        [CDF_HVM; CDF_HAP]
-    | Some "false" when vm_info.hvm ->
-        [CDF_HVM]
-    | Some unknown ->
-        error "VM = %s; Unrecognized value platform/hap=\"%s\"."
-          (Uuid.to_string uuid) unknown ;
-        invalid_arg ("platform/hap=" ^ unknown)
-    | None when vm_info.hvm && vm_info.hap && host_has_hap ->
-        [CDF_HVM; CDF_HAP]
-    | None when vm_info.hvm ->
-        [CDF_HVM]
-    | None ->
-        []
+
+  (* Confirm that the running hypervisor supports a specific capability. *)
+  let assert_capability cap ~on_error =
+    if not (List.mem cap host_info.capabilities) then (
+      let msgstr = on_error () in
+      error "VM = %s: %s" (Uuid.to_string uuid) msgstr ;
+      invalid_arg msgstr
+    )
   in
-  if List.mem CDF_HAP flags && not host_has_hap then (
-    error "VM = %s: Hardware Assisted Paging requested, but not available"
-      (Uuid.to_string uuid) ;
-    invalid_arg "Hardware assisted paging requested but not available"
-  ) ;
-  info
-    "VM = %s; Hardware Assisted Paging will be %s. Use \
-     platform/hap=(true|false) to override"
-    (Uuid.to_string uuid)
-    (if List.mem CDF_HAP flags then "enabled" else "disabled") ;
+
+  (* Xen has two types of virtualisation.  PV "Para-Virtual" which uses ring
+     deprivileging, and HVM "Hardware Virtual Machine" which use Intel VT-x /
+     AMD SVM hardware extensions.  All hybrid guest types (PVHVM, PVH,
+     PVinPVH, etc) are HVM from Xen's point of view. *)
+  let hvm = vm_info.hvm in
+
+  (* Both PV and HVM may be compiled out of Xen.  HVM may be unavailable
+     because of hardware support or firmware/Xen settings. *)
+  assert_capability
+    (if hvm then CAP_HVM else CAP_PV)
+    ~on_error:(fun () ->
+      sprintf "Guest type %s unavailable" (if hvm then "HVM" else "PV")) ;
+
+  (* HVM guests must select a paging mode of either HAP "Hardware Assisted
+     Paging" or Shadow. *)
+  let hap =
+    if not hvm then
+      false
+    else
+      (* If platform/hap=<bool> is explicitly configured, use the setting
+         unconditionally and raise an error if unavailable.  Otherwise, use
+         HAP if available, falling back to Shadow if not. *)
+      let hap =
+        match List.assoc_opt "hap" vm_info.platformdata with
+        | Some "true" ->
+            true
+        | Some "false" ->
+            false
+        | Some unknown ->
+            error "VM = %s; Unrecognized value platform/hap=\"%s\"."
+              (Uuid.to_string uuid) unknown ;
+            invalid_arg ("platform/hap=" ^ unknown)
+        | None ->
+            List.mem CAP_HAP host_info.capabilities
```

Comment:
Alternatively if you want to preserve the flag, then see `generate_create_info`, which has access to a `ty` field, which contains a `hap` field.
You can then set the `vm_info.hap` field based on that.https://github.com/xapi-project/xcp-idl/blob/master/xen/xenops_types.ml#L80-L95

Reply:
That logic would have been ok before 4.12/13 where Xen silently fell back to shadow mode

Reply:
This is initially set in xapi: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_xenops.ml#L367
Perhaps the computation based on the platform flag should be done in XAPI, and in xenopsd we only do the host capability check?
But that means that xapi should do the host capability check too when nothing is specified via a platform field, it gets complicated...

Reply:
It might also be possible to update the existing `hap` field in the HVM record with the desired outcome (and similarly for nested virt). This ends up in the last booted record.

Reply:
Discussed with @robhoes that perhaps a tri-state flag for the XAPI side would be useful, to say what XAPI wants:
* force on - [Some true]
* force off - [Some false] 
* default - [None]

the logic for computing that from platform field would be then in XAPI.

That way if you change the platform field while the VM is running it'll keep using the value from the persistent config on a migration.

And then the outcome of whether to use hap or not would be store somewhere in the persistent_t record (or any of the records that are part of it as suggested above).

Reply:
I think the latest commit here provides the tristate: an absent hap setting in VM platform means use default and otherwise we force it on or off. Obviously if the host doesn't support it then we raise an exception and fail.